### PR TITLE
[core] [swift] Visitor Rules support for Antlr Visitors

### PR DESCRIPTION
**Pull request by lsoncini**
Before submitting a PR, please check that:
 - [ ] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [X] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**

This pull request adds the necessary classes to implement antlr-based abstract rules and generic implementations of both RuleChainVisitor (`AntlrRuleChainVisitor`) and RuleViolationFactory (`AntlrRuleViolationFactory`). 

We also added `AbstractSwiftRule` to serve as an example to show how to implement `AbstractAntlrVisitor` for a specific language and modified the ant-file antlr4.xml to make antlr's autogenerated BaseVisitors extend `AbstractAntlrVisitor`.

Team Raptor.

### CODE REVIEW

**Comment by lsoncini** [comment]
@jsotuyod We are opening this PR in teamraptor's fork in order to avoid chaining too many pull request in PMD as the diffs are growing larger in each one. As soon as `AntlrBaseNode` is merged, we will be rebasing this branch to make reviews easier and opening the corresponding PR.

But in the meantime, we would like to have some review over these classes in order to start making the necessary changes and win some time.

We have some doubts about how PMD uses the `RuleContext` passed as data in each visit for Visitors and jjtAccept or childrenAccept for Nodes. As regards violation registry, we found that giving each rule this data in the start method is enough, but we don't fully know all the scenarios in which this data is used in order to be completely sure this is enough.

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleChainVisitor.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            // visiting by type
+            @Override
+            public Object visit(final AntlrBaseNode node) {
+                indexNode(node);
+                return super.visit(node);
+            }
+        };
+
+        for (int i = 0; i < nodes.size(); i++) {
+            antlrVisitor.visit((AntlrBaseNode) nodes.get(i));
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleChainVisitor.java]: por qué no usamos un foreach en vez de iterar así?

Comment (matifraga) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleChainVisitor.java]: Este método lo tomamos de la implementación de antlr: 
https://github.com/antlr/antlr4/blob/98dc2c0f0249a67b797b151da3adf4ffbc1fd6a1/runtime/Java/src/org/antlr/v4/runtime/tree/AbstractParseTreeVisitor.java#L37

Podríamos agregar un comentario y además modificarlo para usar un foreach, te parece?

Comment (matifraga) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleChainVisitor.java]: Aunque imagino que no vamos a poder reemplazar `node.getChild(i)` en el foreach

**Pull request closed without merging**